typing/types.ml

```diff
@@ -996,19 +1019,135 @@ end
 let eq_type t1 t2 = t1 == t2 || repr t1 == repr t2
 let compare_type t1 t2 = compare (get_id t1) (get_id t2)
 
+(* with-bounds *)
+
+(* Compare types roughly semantically, to allow best-effort deduplication of the types
+   inside of with-bounds.
+
+   This function might compare two types as inequal that are actually equal, but should
+   /never/ compare two types as equal that are not semantically equal. It may go without
+   saying but it also needs to expose a total order.
+
+   Someday, it's probably desirable to merge this, and make it compatible, with
+   [Ctype.eq_type], though that seems quite hard.
+*)
+let rec best_effort_compare_type_expr te1 te2 =
+  if te1 == te2 || repr te1 == repr te2 then 0
+  else
+    let rank ty =
+      match ty.desc with
+      (* Types which must be compared by id *)
+      | Tvar _
+      | Tobject (_, _)
+      | Tfield (_, _, _, _)
+      | Tnil
+      | Tlink _
+      | Tsubst (_, _)
+      | Tvariant _
+      | Tpackage (_, _)
+      | Tarrow (_, _, _, _)
+        -> -ty.id
+      (* Types which we know how to compare structurally*)
+      | Ttuple _ -> 2
+      | Tunboxed_tuple _ -> 3
+      | Tunivar _ -> 4
+      | Tconstr (_, _, _) -> 5
+      | Tpoly (_, _) -> 6
+    in
+    match te1.desc, te2.desc with
+    | Ttuple elts1, Ttuple elts2
+    | Tunboxed_tuple elts1, Tunboxed_tuple elts2 ->
+      List.compare
+        (fun (l1, te1) (l2, te2) ->
+           let l = Option.compare String.compare l1 l2 in
+           if l = 0 then best_effort_compare_type_expr te1 te2 else l
+        )
+        elts1
+        elts2
+    | Tconstr (p1, args1, _), Tconstr (p2, args2, _) ->
+      let p = Path.compare p1 p2 in
+      if p = 0
+      then List.compare best_effort_compare_type_expr args1 args2
+      else p
+    | Tpoly (t1, ts1), Tpoly (t2, ts2) ->
```

Comment:
This case is hard, for a variety of reasons.

- The `ts` are binding sites for variables that may occur in the `t`s. If an in-scope variable is mentioned in a `ts`, it must be shadowed.
- The `ts` are set-like, in that order doesn't matter.
- Even worse, if a `ts` binds a variable that is then unused in a `t`, the `Tpoly` has no effect.
- If the `ts` is empty (which happens plenty of times in practice), then the `Tpoly` has no effect.

Maybe I'm missing some, too. All that said, given the "best effort" nature, maybe this is correct? At the least, there should be a loud comment saying this is mostly broken.